### PR TITLE
Add support for react-native

### PR DESCRIPTION
**Pull request by giautm**
I didn't tested on web, please test before merge.

### CODE REVIEW

**Comment by giautm** [comment]
@jaydenseric: I tested it on react-native, `react-native-apollo-upload-client` is this fork and it work on my project.

Please test again on browser.

**Comment by jaydenseric** [comment]
A great start! Thanks for rebasing it on the newer version.

File: package.json [hunk 1]
```diff
@@ -22,20 +22,23 @@
   ],
   "dependencies": {
     "apollo-client": "^1.4.2",
-    "babel-runtime": "^6.23.0"
+    "babel-preset-env": "^1.5.2",
```
Comment (jaydenseric) [inline package.json]: These dependencies should be moved back to dev dependencies.

File: rollup.config.js [hunk 2]
```diff
@@ -7,8 +9,11 @@ export default {
   external: Object.keys(pkg.dependencies),
   plugins: [
     babel({
-      runtimeHelpers: true
-    })
+      plugins: ['external-helpers'],
+      externalHelpers: true
+    }),
+    nodeResolve(),
+    commonjs()
```
Comment (jaydenseric) [inline rollup.config.js]: Can you help me understand why these changes are necessary?

Comment (giautm) [inline rollup.config.js]: I need it for `import Symbol from 'es6-symbol/polyfill'`. If not, the transformer tricked as external

File: package.json (hunk 1 above)
Comment (giautm) [inline package.json]: No, if move them back to dev, i need to install all of them to build on react-native.

**Comment by giautm** [comment]
@jaydenseric : Can you setup a minium server somewhere (heroku) so i can submit a runable example? (I will use Expo).

**Comment by jaydenseric** [comment]
@giautm You can test new versions with this example project: https://github.com/jaydenseric/apollo-upload-examples.

**Comment by giautm** [comment]
Yah, I will write an example of react-native at tonight. 👍

**Comment by giautm** [comment]
I trying to replace `apollo-upload-client` with `react-native-apollo-upload-client` in `apollo-upload-examples/app` then got following error:
```
ReferenceError: babelHelpers is not defined
    at /Users/admin/Devel/giautm/apollo-upload-examples/app/node_modules/react-native-apollo-upload-client/dist/apollo-upload-client.js:408:3
    at Object.<anonymous> (/Users/admin/Devel/giautm/apollo-upload-examples/app/node_modules/react-native-apollo-upload-client/dist/apollo-upload-client.js:455:2)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/admin/Devel/giautm/apollo-upload-examples/app/.next/dist/helpers/with-data.js:47:38)
```
How we can make a build work with both react-native and browser? :(

File: package.json (hunk 1 above)
Comment (jaydenseric) [inline package.json]: I still don't understand why, because the dist files are published on npm. having them in dependencies would mean they are redundantly installed in production environments.

File: rollup.config.js (hunk 2 above)
Comment (jaydenseric) [inline rollup.config.js]: But it is external? Other imports were working ok.

File: package.json (hunk 1 above)
Comment (giautm) [inline package.json]: No, your dist files not bundled with there dependencies. See the screen shot, because your build with runtime-transfer.
screen shot 2017-06-22 at 12 44 29

Comment (jaydenseric) [inline package.json]: But thats how it's meant to work. So that when you create an app bundle, apollo-upload-client does include baked-in helper code, duplicated by other dependencies – they get shared between modules and only appear once in your final Webpack bundle. Also, babel does not run on node_modules, only the current module.

File: src/file-input.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/**
+ * @flow
+ */
+// On Android we need a polyfill for Symbol
+import Symbol from 'es6-symbol/polyfill'
```
Comment (jaydenseric) [inline src/file-input.js]: I havn't used symbols before, but is it really necessary to use them here? This polyfill looks to be as big as `apollo-upload-client` itself.

File: package.json (hunk 1 above)
Comment (giautm) [inline package.json]: yah, i will try again with `transform-runtime`, i will that report error if happen.

File: src/network-interface.js [hunk 4]
```diff
@@ -4,7 +4,10 @@ import { extractRequestFiles } from './helpers'
 export class HTTPUploadNetworkInterface extends HTTPFetchNetworkInterface {
   fetchFromRemoteEndpoint({ request, options }) {
     // Skip upload proccess if SSR
-    if (typeof window !== 'undefined') {
+    if (
+      typeof window !== 'undefined' ||
+      (typeof global !== 'undefined' && typeof global.FormData !== 'undefined')
+    ) {
```
Comment (jaydenseric) [inline src/network-interface.js]: Interesting idea to just look for the existence of FormData in global scope. Maybe we could replace the whole condition with just `if (FormData) {`

Comment (giautm) [inline src/network-interface.js]: No if just `if (FormData) {` wil cause a `ReferenceError: FormData is not defined`

**Comment by giautm** [comment]
@jaydenseric : I reverted babel config and bump `react-native-apollo-upload-client` and test on react-native. Still need to keep there module at `dependencies`:
```
    "babel-plugin-transform-runtime": "^6.23.0",
    "babel-preset-env": "^1.5.2",
    "babel-preset-stage-0": "^6.24.1",
```
image

File: src/file-input.js (hunk 3 above)
Comment (giautm) [inline src/file-input.js]: Only need on Android, babel wil bundle if need.

**Comment by jaydenseric** [comment]
@giautm I see that you have published your fork to npm under [`react-native-apollo-upload-client`](https://www.npmjs.com/search?q=react-native-apollo-upload-client) without attribution, in violation of my MIT licence. I would appreciate it if you would unpublish it.

Please be patient, I have a `support-react-native` branch I will be publishing shortly.

**Comment by giautm** [comment]
@jaydenseric : Sorry, that was accident. I merged wrong branch to master. I just revert it. :(

**Comment by giautm** [comment]
I can't unpublish `react-native-apollo-upload-client` right now because: http://blog.npmjs.org/post/141905368000/changes-to-npms-unpublish-policy. Sorry about that. :(

**Comment by jaydenseric** [comment]
@giautm thanks for your work, the way forward can be seen [here](https://github.com/jaydenseric/apollo-upload-client/issues/10#issuecomment-310314433).
